techniques/android/MASTG-TECH-0017.md

```diff
@@ -3,50 +3,30 @@ title: Decompiling Java Code
 
-If you want to look directly into Java source code on a GUI, simply open your APK using @MASTG-TOOL-0018 or @MASTG-TOOL-0014.
+While DEX bytecode and Java bytecode are not the same, they can be converted into each other. As a result, Android applications can be decompiled using Java decompilers by first converting the `.dex` files into `.jar` files and then decompiling the `.jar` files. Since one application can have multiple `.dex`, this is a tedious process, and luckily many decompilers have native support for `.apk` files, thereby taking care of this process for you. Some decompilers also work directly on DEX bytecode, rather than converting it into Java bytecode first.
 
-Android decompilers go one step further and attempt to convert Android bytecode back into Java source code, making it more human-readable. Fortunately, Java decompilers generally handle Android bytecode well. The above mentioned tools embed, and sometimes even combine, popular free decompilers such as:
+Due to the long legacy of Java applications, many decompilers can be used to decompile Android applications. While some very strong commercial Android decompilers exist, there are free decompilers which rival commercial decompilers in many aspects. The most popular free decompiler is @MASTG-TOOL-0018, which is actively developed. If jadx fails to decompile a part of the code, the easiest alternative is @MASTG-TOOL-0014, which combines six different decompilers in one application (DJ-GUI, Procyon, CFR, Fernflower, Krakatau and JADX-Core). Both of these decompilers have native support for `.apk` files, and while it may seem that bytecodeviewer is the better choice, jadx has many more UI features that offer a much more pleasant user experience than bytecodeviewer.
 
-- [JD](http://java-decompiler.github.io/ "JD")
-- [JAD](http://www.javadecompilers.com/jad "JAD")
-- @MASTG-TOOL-0018
-- [Procyon](https://github.com/mstrobel/procyon "Procyon")
-- [CFR](https://www.benf.org/other/cfr/ "CFR")
+!!! warning
 
-Alternatively you can use the @MASTG-TOOL-0010 extension for Visual Studio Code or run @MASTG-TOOL-0012 on your APK or use the exported files from the previous tools to open the reversed source code on your preferred IDE.
+    Decompilation can always fail, either due to intentional manipulation of the `.dex` files, or simply due to bugs in the decompilers. If all decompilers fail, you can always fall back to @MASTG-TECH-0016.
 
-In the following example we'll be using @MASTG-APP-0003. First, let's install the app on a device or emulator and run it to see what the crackme is about.
+Let's take a look at the decompiled version of @MASTG-APP-0003 in a few different decompilers:
 
-<img src="Images/Chapters/0x05c/crackme-1.png" width="400px" />
+## Using jadx-gui
```

Comment:
add `## Using jadx` and highlight limitations, e.g. no built-in "find usage" search, and more. And point to the solution, e.g. use grep/ripgrep or jadx-gui, etc.

Reply:
Actually, we have TECHs for those things as well, see for example:

https://mas.owasp.org/MASTG/techniques/android/MASTG-TECH-0020/

https://mas.owasp.org/MASTG/techniques/android/MASTG-TECH-0019/

https://mas.owasp.org/MASTG/techniques/android/MASTG-TECH-0021/

So it'd be nice to link to them and also update them (that can happen on a separate PR of course)